hack/cleanup-ag: spurious affinity group cleaner

rohityadavcloud opened the pull request:
*Issue #, if available:*

Related to https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack/issues/240

*Description of changes:*
This cleanup tool helps to remove CloudStack affinity groups from CAPC management cluster which are not assigned to any instances, which may have been created as a side effect of other operations. This tool checks all the cloudstackaffinitygroups using its CloudStack cluster specific credential(s) and uses cmk to check if the affinity group have no instances assigned. In dry-run, it outputs such affinity groups otherwise it deletes them.

*Testing performed:*

- Created a CAPC cluster with affinity groups specified as `anti` or `pro`
```
spec:
  template:
    spec:
      affinity: anti
      offering:
        name: Large
      template:
        name: rockylinux-8-kube-v1.27.2-kvm
```
- Delete the machine deployment to simulate upgrade or similar operations
```
kubectl delete  machinedeployment capc-cluster-md-0
```
- Re-applied the capc-cluster yaml, to see new machines being deployed and new affinity groups created
- However, older affinity groups aren't cleaned up
- Install dependencies for this tool such as cmk and jq (note: cmk should be latest from main branch or get this pre-rc build: https://github.com/apache/cloudstack-cloudmonkey/releases/tag/6.4.0-rc1)
- Run this tool in dry-run mode:
```
./cleanup-ag.sh -d
Checking CloudStack Affinity Group: antiaffinity-capc-cluster-control-plane-746e19e9-ea06-4ca3-b9b1-ebeb4753e6c7-failure-domain-1
Checking CloudStack Affinity Group: antiaffinity-capc-cluster-md-0-7f8d8bf7c8x946qk-c7a7a6d9-7d3d-480d-8d46-ff8260b8cd7f-failure-domain-1
Checking CloudStack Affinity Group: antiaffinity-capc-cluster-md-0-7f8d8bf7c8xfph69-915f82e5-e346-4324-8dd6-b793ea6768e5-failure-domain-1
Found Affinity Group (4db6e60d-942d-48b9-9e8d-f27a911ceaf4) with no instances assigned: antiaffinity-capc-cluster-md-0-7f8d8bf7c8xfph69-915f82e5-e346-4324-8dd6-b793ea6768e5-failure-domain-1
[DRY RUN] Affinity Group (4db6e60d-942d-48b9-9e8d-f27a911ceaf4) antiaffinity-capc-cluster-md-0-7f8d8bf7c8xfph69-915f82e5-e346-4324-8dd6-b793ea6768e5-failure-domain-1 has been removed
Checking CloudStack Affinity Group: proaffinity-capc-cluster2-control-plane-f3cab1ca-32a0-4a41-ad2f-a681bd5a60c8-failure-domain-1
Checking CloudStack Affinity Group: proaffinity-capc-cluster2-md-0-74f9cbf9dx445dq-df4594c2-9b98-44ab-8901-acfadf68445e-failure-domain-1
Found Affinity Group (a7ea7e9f-b7c8-4607-bc15-3ebe9bc95594) with no instances assigned: proaffinity-capc-cluster2-md-0-74f9cbf9dx445dq-df4594c2-9b98-44ab-8901-acfadf68445e-failure-domain-1
[DRY RUN] Affinity Group (a7ea7e9f-b7c8-4607-bc15-3ebe9bc95594) proaffinity-capc-cluster2-md-0-74f9cbf9dx445dq-df4594c2-9b98-44ab-8901-acfadf68445e-failure-domain-1 has been removed
Checking CloudStack Affinity Group: proaffinity-capc-cluster2-md-0-74f9cbf9dxdqk85-160cd89f-4361-4db4-aaf0-e93c28a0dae0-failure-domain-1
Checking CloudStack Affinity Group: proaffinity-capc-cluster2-md-0-74f9cbf9dxlxx6c-bf59c3c6-bc8d-44e6-917f-ececf69e75bb-failure-domain-1
Found Affinity Group (c800d58c-31a8-493c-b45e-94a8e9f5a149) with no instances assigned: proaffinity-capc-cluster2-md-0-74f9cbf9dxlxx6c-bf59c3c6-bc8d-44e6-917f-ececf69e75bb-failure-domain-1
[DRY RUN] Affinity Group (c800d58c-31a8-493c-b45e-94a8e9f5a149) proaffinity-capc-cluster2-md-0-74f9cbf9dxlxx6c-bf59c3c6-bc8d-44e6-917f-ececf69e75bb-failure-domain-1 has been removed
```
- Next, run the tool to cleanup:
```
> ./cleanup-ag.sh
Checking CloudStack Affinity Group: antiaffinity-capc-cluster-control-plane-746e19e9-ea06-4ca3-b9b1-ebeb4753e6c7-failure-domain-1
Checking CloudStack Affinity Group: antiaffinity-capc-cluster-md-0-7f8d8bf7c8x946qk-c7a7a6d9-7d3d-480d-8d46-ff8260b8cd7f-failure-domain-1
Checking CloudStack Affinity Group: antiaffinity-capc-cluster-md-0-7f8d8bf7c8xfph69-915f82e5-e346-4324-8dd6-b793ea6768e5-failure-domain-1
Found Affinity Group (4db6e60d-942d-48b9-9e8d-f27a911ceaf4) with no instances assigned: antiaffinity-capc-cluster-md-0-7f8d8bf7c8xfph69-915f82e5-e346-4324-8dd6-b793ea6768e5-failure-domain-1
cloudstackaffinitygroup.infrastructure.cluster.x-k8s.io "antiaffinity-capc-cluster-md-0-7f8d8bf7c8xfph69-915f82e5-e346-4324-8dd6-b793ea6768e5-failure-domain-1" deleted
Affinity Group (4db6e60d-942d-48b9-9e8d-f27a911ceaf4) antiaffinity-capc-cluster-md-0-7f8d8bf7c8xfph69-915f82e5-e346-4324-8dd6-b793ea6768e5-failure-domain-1 has been removed
Checking CloudStack Affinity Group: proaffinity-capc-cluster2-control-plane-f3cab1ca-32a0-4a41-ad2f-a681bd5a60c8-failure-domain-1
Checking CloudStack Affinity Group: proaffinity-capc-cluster2-md-0-74f9cbf9dx445dq-df4594c2-9b98-44ab-8901-acfadf68445e-failure-domain-1
Found Affinity Group (a7ea7e9f-b7c8-4607-bc15-3ebe9bc95594) with no instances assigned: proaffinity-capc-cluster2-md-0-74f9cbf9dx445dq-df4594c2-9b98-44ab-8901-acfadf68445e-failure-domain-1
cloudstackaffinitygroup.infrastructure.cluster.x-k8s.io "proaffinity-capc-cluster2-md-0-74f9cbf9dx445dq-df4594c2-9b98-44ab-8901-acfadf68445e-failure-domain-1" deleted
Affinity Group (a7ea7e9f-b7c8-4607-bc15-3ebe9bc95594) proaffinity-capc-cluster2-md-0-74f9cbf9dx445dq-df4594c2-9b98-44ab-8901-acfadf68445e-failure-domain-1 has been removed
Checking CloudStack Affinity Group: proaffinity-capc-cluster2-md-0-74f9cbf9dxdqk85-160cd89f-4361-4db4-aaf0-e93c28a0dae0-failure-domain-1
Checking CloudStack Affinity Group: proaffinity-capc-cluster2-md-0-74f9cbf9dxlxx6c-bf59c3c6-bc8d-44e6-917f-ececf69e75bb-failure-domain-1
Found Affinity Group (c800d58c-31a8-493c-b45e-94a8e9f5a149) with no instances assigned: proaffinity-capc-cluster2-md-0-74f9cbf9dxlxx6c-bf59c3c6-bc8d-44e6-917f-ececf69e75bb-failure-domain-1
cloudstackaffinitygroup.infrastructure.cluster.x-k8s.io "proaffinity-capc-cluster2-md-0-74f9cbf9dxlxx6c-bf59c3c6-bc8d-44e6-917f-ececf69e75bb-failure-domain-1" deleted
Affinity Group (c800d58c-31a8-493c-b45e-94a8e9f5a149) proaffinity-capc-cluster2-md-0-74f9cbf9dxlxx6c-bf59c3c6-bc8d-44e6-917f-ececf69e75bb-failure-domain-1 has been removed
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

<!-- If this is a security issue, please do not discuss on GitHub. Please report any suspected or confirmed security issues to AWS Security https://aws.amazon.com/security/vulnerability-reporting/ -->

rohityadavcloud commented:
/assign @vishesh92

cc @kiranchavala

weizhouapache commented on hack/cleanup-ag.sh:
```diff
@@ -0,0 +1,124 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires: jq, kubectl, cmk (get https://github.com/apache/cloudstack-cloudmonkey/releases/tag/6.4.0-rc1 or later)
```
@rohityadavcloud 
would be good to check if cmk is installed ?

rohityadavcloud replied:
done

vishesh92 commented on hack/cleanup-ag.sh:
```diff
@@ -0,0 +1,124 @@
+    cluster=$(get_cluster $ag)
+    for credential in $(get_cluster_credentials $cluster); do
+      setup_acs_credentials $credential
+      CS_AG_ID=$(kubectl get cloudstackaffinitygroup $ag -o json | jq -r '.spec.id')
```
Specify namespace or use -A to get the id here.

rohityadavcloud replied:
Same as above:
```
 > kubectl get cloudstackaffinitygroup antiaffinity-capc-cluster-control-plane-746e19e9-ea06-4ca3-b9b1-ebeb4753e6c7-failure-domain-1 -A
error: a resource cannot be retrieved by name across all namespaces
```

vishesh92 commented on hack/cleanup-ag.sh:
```diff
@@ -0,0 +1,124 @@
+
+get_cluster_credentials() {
+  cluster=$1
+  kubectl get cloudstackcluster $cluster -o json | jq -r '.spec.failureDomains[].acsEndpoint.name' | uniq
```
Add -A or specify namespace.

rohityadavcloud replied:
@vishesh92 gives me an error - note we're getting a specific cluster by name:
```
> kubectl get cloudstackcluster capc-cluster -A
error: a resource cannot be retrieved by name across all namespaces
```

vishesh92 commented on hack/cleanup-ag.sh:
```diff
@@ -0,0 +1,124 @@
+
+get_cluster() {
+  affinityGroup=$1
+  kubectl get cloudstackaffinitygroup $affinityGroup -o json | jq -r '.metadata.labels."cluster.x-k8s.io/cluster-name"'
```
Add -A or specify namespace.

rohityadavcloud commented on hack/cleanup-ag.sh:
```diff
@@ -0,0 +1,124 @@
+
+setup_acs_credentials() {
+  credential=$1
+  export CS_URL=$(kubectl get secret $credential -o json | jq -r '.data."api-url"' | base64 -D)
```
same as above: here we're getting a specific credential by name
```
> kubectl get secret cloudstack-credentials -A
error: a resource cannot be retrieved by name across all namespaces
```

rohityadavcloud commented:
Thanks @weizhouapache @vishesh92 - incorporated your comments, pl re-review.

rohityadavcloud commented:
/hold

Holding for manual QA cc @kiranchavala

chrisdoherty4 commented on hack/cleanup-ag.sh:
`ag` isn't, in my opinion, a common acronym. Would you mind expanding to `clean-affinity-groups.sh`?

rohityadavcloud replied:
Fixed

chrisdoherty4 commented on hack/cleanup-ag.sh:
```diff
@@ -0,0 +1,154 @@
+set -o nounset
+set -o pipefail
+
+export DRY_RUN=false
+export VERBOSE=false
+export NAMESPACE=default
+export KUBECONFIG=$HOME/.kube/config
```
Are _all_ of these consumed by child processes? If not, we can drop the `export`.

rohityadavcloud replied:
Thanks, checked all and fixed

chrisdoherty4 commented on hack/cleanup-ag.sh:
```diff
@@ -0,0 +1,154 @@
+}
+
+get_cluster_credentials() {
+  cluster=$1
```
```suggestion
  local cluster=$1
```

chrisdoherty4 commented on hack/cleanup-ag.sh:
```diff
@@ -0,0 +1,154 @@
+}
+
+setup_acs_credentials() {
+  credential=$1
```
```suggestion
  local credential=$1
```

chrisdoherty4 commented on hack/cleanup-ag.sh:
```diff
@@ -0,0 +1,154 @@
+main() {
+  for ag in $(get_affinity_groups); do
+    echo "[info] Checking CloudStack Affinity Group: $ag"
+    cluster=$(get_cluster $ag)
```
```suggestion
    local cluster=$(get_cluster $ag)
```

chrisdoherty4 commented on hack/cleanup-ag.sh:
```diff
@@ -0,0 +1,154 @@
+    cluster=$(get_cluster $ag)
+    for credential in $(get_cluster_credentials $cluster); do
+      setup_acs_credentials $credential
+      CS_AG_ID=$(_kubectl get cloudstackaffinitygroup $ag | jq -r '.spec.id')
+      CS_AG_VMS=$(_cmk list affinitygroups id=$CS_AG_ID | jq -r '.affinitygroup[0].virtualmachineIds')
```
```suggestion
      local CS_AG_ID=$(_kubectl get cloudstackaffinitygroup $ag | jq -r '.spec.id')
      local CS_AG_VMS=$(_cmk list affinitygroups id=$CS_AG_ID | jq -r '.affinitygroup[0].virtualmachineIds')
```
?

rohityadavcloud commented:
Thanks @chrisdoherty4 I've addressed the comments;

Sample output with dryrun and verbose flag enabled:
```bash
> ./cleanup-affinity-groups.sh -d  -v
[debug] [options] DRY_RUN=true
[debug] [options] VERBOSE=true
[debug] [options] NAMESPACE=default
[debug] [options] KUBECONFIG=/Users/rohit/.kube/config
[info] Checking CloudStack Affinity Group: antiaffinity-capc-cluster-control-plane-746e19e9-ea06-4ca3-b9b1-ebeb4753e6c7-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Checking CloudStack Affinity Group: antiaffinity-capc-cluster-md-0-7f8d8bf7c8xbgqhx-81ca936d-797e-447f-88da-8ba04624eb44-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Found Affinity Group (2a64d1cb-f4db-48b4-b5c7-803341497cb7) with no instances assigned: antiaffinity-capc-cluster-md-0-7f8d8bf7c8xbgqhx-81ca936d-797e-447f-88da-8ba04624eb44-failure-domain-1
[info] [dryrun] Affinity Group (2a64d1cb-f4db-48b4-b5c7-803341497cb7) antiaffinity-capc-cluster-md-0-7f8d8bf7c8xbgqhx-81ca936d-797e-447f-88da-8ba04624eb44-failure-domain-1 has been removed
[info] Checking CloudStack Affinity Group: antiaffinity-capc-cluster-md-0-7f8d8bf7c8xn6z9g-0c9f91f1-18f9-4767-b861-24c72ea2fe68-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Checking CloudStack Affinity Group: proaffinity-capc-cluster2-control-plane-f3cab1ca-32a0-4a41-ad2f-a681bd5a60c8-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Checking CloudStack Affinity Group: proaffinity-capc-cluster2-md-0-74f9cbf9dxd8x6f-50e535ef-e808-4bb9-85a6-cd76e287f7fc-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Checking CloudStack Affinity Group: proaffinity-capc-cluster2-md-0-74f9cbf9dxdqk85-160cd89f-4361-4db4-aaf0-e93c28a0dae0-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Found Affinity Group (5cd49b3d-05d7-45c7-8ffe-ce34640e6c4c) with no instances assigned: proaffinity-capc-cluster2-md-0-74f9cbf9dxdqk85-160cd89f-4361-4db4-aaf0-e93c28a0dae0-failure-domain-1
[info] [dryrun] Affinity Group (5cd49b3d-05d7-45c7-8ffe-ce34640e6c4c) proaffinity-capc-cluster2-md-0-74f9cbf9dxdqk85-160cd89f-4361-4db4-aaf0-e93c28a0dae0-failure-domain-1 has been removed
```

Sample output (without dryrun):
```bash
./cleanup-affinity-groups.sh -v
[debug] [options] DRY_RUN=false
[debug] [options] VERBOSE=true
[debug] [options] NAMESPACE=default
[debug] [options] KUBECONFIG=/Users/rohit/.kube/config
[info] Checking CloudStack Affinity Group: antiaffinity-capc-cluster-control-plane-746e19e9-ea06-4ca3-b9b1-ebeb4753e6c7-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Checking CloudStack Affinity Group: antiaffinity-capc-cluster-md-0-7f8d8bf7c8xbgqhx-81ca936d-797e-447f-88da-8ba04624eb44-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Found Affinity Group (2a64d1cb-f4db-48b4-b5c7-803341497cb7) with no instances assigned: antiaffinity-capc-cluster-md-0-7f8d8bf7c8xbgqhx-81ca936d-797e-447f-88da-8ba04624eb44-failure-domain-1
cloudstackaffinitygroup.infrastructure.cluster.x-k8s.io "antiaffinity-capc-cluster-md-0-7f8d8bf7c8xbgqhx-81ca936d-797e-447f-88da-8ba04624eb44-failure-domain-1" deleted
[info] Affinity Group (2a64d1cb-f4db-48b4-b5c7-803341497cb7) antiaffinity-capc-cluster-md-0-7f8d8bf7c8xbgqhx-81ca936d-797e-447f-88da-8ba04624eb44-failure-domain-1 has been removed
[info] Checking CloudStack Affinity Group: antiaffinity-capc-cluster-md-0-7f8d8bf7c8xn6z9g-0c9f91f1-18f9-4767-b861-24c72ea2fe68-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Checking CloudStack Affinity Group: proaffinity-capc-cluster2-control-plane-f3cab1ca-32a0-4a41-ad2f-a681bd5a60c8-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Checking CloudStack Affinity Group: proaffinity-capc-cluster2-md-0-74f9cbf9dxd8x6f-50e535ef-e808-4bb9-85a6-cd76e287f7fc-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Checking CloudStack Affinity Group: proaffinity-capc-cluster2-md-0-74f9cbf9dxdqk85-160cd89f-4361-4db4-aaf0-e93c28a0dae0-failure-domain-1
[debug] Using CloudStack Control Plane URL: http://192.168.1.10:8080/client/api and CloudStack Account: capc and User: capc
[info] Found Affinity Group (5cd49b3d-05d7-45c7-8ffe-ce34640e6c4c) with no instances assigned: proaffinity-capc-cluster2-md-0-74f9cbf9dxdqk85-160cd89f-4361-4db4-aaf0-e93c28a0dae0-failure-domain-1
cloudstackaffinitygroup.infrastructure.cluster.x-k8s.io "proaffinity-capc-cluster2-md-0-74f9cbf9dxdqk85-160cd89f-4361-4db4-aaf0-e93c28a0dae0-failure-domain-1" deleted
[info] Affinity Group (5cd49b3d-05d7-45c7-8ffe-ce34640e6c4c) proaffinity-capc-cluster2-md-0-74f9cbf9dxdqk85-160cd89f-4361-4db4-aaf0-e93c28a0dae0-failure-domain-1 has been removed
```

chrisdoherty4 commented:
/lgtm

weizhouapache reviewed (approved):
lgtm

vishesh92 commented:
lgtm

kiranchavala reviewed (approved):
/lgtm

Tested the tool manually and its working fine with all options 
```

./cleanup-affinity-groups.sh -d
./cleanup-affinity-groups.sh -v
./cleanup-affinity-groups.sh 
./cleanup-affinity-groups.sh -n default
./cleanup-affinity-groups.sh -n default -v -d
./cleanup-affinity-groups.sh -n default -d
```

rohityadavcloud commented:
/unhold

rohityadavcloud commented:
/approve
/lgtm
